### PR TITLE
Remove resolver=2 from Cargo.toml and add it to the Windows build

### LINKED ISSUE #22603: getBlock spike triggers permanent catchup difficulty by rayon/crossbeam bug
#### Problem

> also, i finally spotted a pretty bad bug in our deep-down crate (rayon => crossbeam), affecting rpc usage: https://github.com/crossbeam-rs/crossbeam/blob/cd453946becccd7e5b29f7305e0d45e8f6417347/crossbeam-epoch/src/internal.rs#L390 (in short, we need to bump this number like by x128 or needs proper adaptive amortization mechanism)
> 
> this is the root cause why blockstore-related rpc thread pool magically affects blockstore_processor thread pool, which has been mysteries for months. also this explains another odd thing that the effect persists even after rpc req burst.
also, this performance degradation is proportional to the system core count. irony here is that more cpu cores for rpc => more slowdown/difficulty to keep catchup to the cluster.
for the moment, i think i gotta focus on stake credit_observed auto rewind. so, please ping me if anyone want to take this one (to ship more sooner)
> 
> ps: so rocksdb is innocent this time too. lol This also explains the SOLANA_RAYON_THREADS=8 hack works. 

#### technical mechanism of the problem

in short, rayon/crossbeam isn't written with the assumption of idling threads in various pools, which exceeds the system's core count by far. and we're doing _that_ unfortunately.

firstly, rayon thread pool is backed by `crossbeam-deque`, in turn `crossbeam-epoch` (a garbage collection library). and each rayon worker thread periodically runs the gc bookeeping code.

and the bookkeeping code isn't well optimized for our usecase. it contains a periodic linear scan of (lockless) linked list of _all rayon worker threads in the os process_ as a part of it. so, as the number of threads increases, each scanning process takes longer, burdening every rayon threads. and the `PINNINGS_BETWEEN_COLLECT` constant controls the frequency and it's too short (i.e. frequent) for us. so, seemingly independent thread pools affects each other. (that was the mystery part)

and solana-validator's rpc subsystem spawns new set of rayon thread pool if its blockstore related jsonrpc method is invoked.
that causes sudden increase of alive rayon threads, so the crossbeam's scanning is increased. finally, if a certain threshold is crossed depending on the system configuration, the catchup (the replay stage) related thread pools (`blockstore_processor` and `entry` ones) are too wasting cycles in the gc bookkeeping to process transactions/entries in timely manner.

#### Proposed Solution

- fix the upstream (i think it will take some effort for creating demo code and the actual fix to be merged there)
- monkey-patch the crossbeam-deque. (as a short term fix if the upstream method takes way long time)

### PR DESCRIPTION
**Pull request by willhickey**
See discussion on PR #26555

#### Problem
PR #24196 added `resolver = "2"` to Cargo.toml. That fixed the broken Windows build but is causing other problems (see https://github.com/solana-labs/solana/issues/22603)

#### Summary of Changes
* Remove `resolver = "2"` from Cargo.toml
* Add steps to the Windows build that add `resolver = "2"` to the local copy of Cargo.toml immediately before the build.

Fixes #22603

### CODE REVIEW

**Review by brennanwatt — APPROVED** [review]
LGTM

File: .github/workflows/release-artifacts.yml [hunk 1]
```diff
@@ -54,6 +60,12 @@ jobs:
           echo "::set-output name=tag::$CI_TAG"
           eval "$(ci/channel-info.sh)"
           echo "::set-output name=channel::$CHANNEL"
+          echo "resolver = \"2\"" >> Cargo.toml
+          echo ${{ secrets.GITHUB_TOKEN }} | gh auth login --with-token
+          git config --global user.email "<>"
+          git config --global user.name "Github Action Bot"
+          git add -u
+          git commit -m "Add resolver=2"
```
[the remark this answers is not in the document]
Comment (willhickey) [reply .github/workflows/release-artifacts.yml]: Yeah good catch. I moved the logic into ci/publish-tarball.sh so it's after the `git reset --hard`

**Review by yihau — APPROVED** [review]
👍

**Comment by ryoqun** [comment]
@willhickey @yihau how about backporting this to v1.10 branch? do you see any concerns? this pr has been in for awhile in master and causing no problem? as a context, I'm intending to ship #26555 to v1.10.

**Comment by yihau** [comment]
just installed an edge bin on windows. I think it works fine.

**Comment by ryoqun** [comment]
@willhickey @yihau oops. it looks like this pr isn't needed for #26555 actually. sorry. I'm reverting this.

I was seeing odd error because, `resolver = "2"` is part of the newly added `[patch]` section. this is a foot gun...

i.e. this works:

```patch
 # This prevents a Travis CI error when building for Windows.
 resolver = "2"
+
+# for details, see https://github.com/solana-labs/crossbeam/commit/fd279d707025f0e60951e429bf778b4813d1b6bf
+[patch.crates-io]
+crossbeam-epoch = { git = "https://github.com/solana-labs/crossbeam", rev = "fd279d707025f0e60951e429bf778b4813d1b6bf" 
```

this won't:

```patch
+
+# for details, see https://github.com/solana-labs/crossbeam/commit/fd279d707025f0e60951e429bf778b4813d1b6bf
+[patch.crates-io]
+crossbeam-epoch = { git = "https://github.com/solana-labs/crossbeam", rev = "fd279d707025f0e60951e429bf778b4813d1b6bf" 

 # This prevents a Travis CI error when building for Windows. <= Even this comment doesn't terminate the preceeding patch section parsing context...
 resolver = "2"  # <= this is now meaning to patch a non-existing crate called `resolver`... I'm dumb
```

```
$ cargo clippy
    Updating crates.io index
error: failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  patch for `resolver` in `https://github.com/rust-lang/crates.io-index` failed to resolve

Caused by:
  The patch location `registry `crates-io`` contains a `resolver` package with versions `0.1.0, 0.1.1`, but the patch definition requires `^2`.
  Check that the version in the patch location is what you expect, and update the patch definition to match.
```

**Pull request merged**